File: 雷达仿真/知识星球/6种常见通信干扰信号/widenoiseJam.m
% produce a fragment of narrow band noise 
clear;close all;jnr = 20; % jam-noise ratio
Fs = 2000;                    % Sampling frequency
T = 1/Fs;                     % Sample time
L = 2*Fs;                     % Length of signal
t = (0:L-1)*T;                % Time vector
P=1;% power of noise
y = wgn(L,1,P); % white noise
% 
% y = awgn(x,jnr,'measured');    % Sinusoids plus noise
figure;plot(1000*t(1:L/2),y(1:L/2));
title('White Noise')
xlabel('time (milliseconds)');grid on;

NFFT = 2^nextpow2(L); % Next power of 2 from length of y
Y = fft(y,NFFT)/L;
f = Fs/2*linspace(0,1,NFFT/2+1);
% Plot single-sided amplitude spectrum.
figure;plot(f,2*10*log10(abs(Y(1:NFFT/2+1)))) 
title('Single-Sided Amplitude Spectrum of White Noise')
xlabel('Frequency (Hz)');ylabel('|Y(f)|(dB)');grid on;
% ylim([10e-4,1])

%% bandpass filter
WI = 50; % window of narrow band
FJ = 600;% jamming frequency
fl_kaiser = [FJ-WI/2 FJ-WI/4 FJ+WI/4 FJ+WI/2];
fl_mag = [0 1 0];
fl_dev = [0.05 0.05 0.05];
[fl_n_kaiser,fl_wn,fl_beta,fl_ftype]=kaiserord(fl_kaiser,fl_mag,fl_dev,Fs);
h= fir1(fl_n_kaiser,fl_wn,fl_ftype,kaiser(fl_n_kaiser+1,fl_beta));
Y_bp = filter(h,1,y);% the result of filter in time domain
figure;
freqz(h)
figure;
subplot(2,1,1);plot(1000*t(1:L/2),y(1:L/2));% original noise
title('White Noise');xlabel('time (milliseconds)');grid on;
subplot(2,1,2);plot(1000*t(1:L/2),Y_bp(1:L/2));% original noise
title('Narrow Band White Noise');xlabel('time (milliseconds)');grid on;

figure;set(gcf,'Color','w','Position',[400 300 600 300]);
plot(1000*t(1:L/2),Y_bp(1:L/2),'Color',[0,0.4,0.8]);grid on;
title('Narrowband Noise jam')
xlabel('time (milliseconds)');


NFFT = 2^nextpow2(L); % Next power of 2 from length of y
Y = fft(Y_bp,NFFT)/L;
f = Fs/2*linspace(0,1,NFFT/2+1);
% Plot single-sided amplitude spectrum.
figure;set(gcf,'Color','w','Position',[400 300 600 300]);
plot(f,2*10*log10(abs(Y(1:NFFT/2+1))),'Color',[0,0.4,0.8]);grid on;
title('Single-Sided Amplitude Spectrum of Narrowband Noise jam')
xlabel('Frequency (Hz)');ylabel('|Y(f)|(dB)');grid on;
% ylim([10e-4,1])

File: 雷达仿真/知识星球/6种常见通信干扰信号/widenoiseJam.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Oct 26 14:47:27 2025

@author: jack
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy import signal
from scipy.signal import kaiserord, firwin, lfilter, freqz
import warnings
warnings.filterwarnings('ignore')

# 产生窄带噪声片段
jnr = 20  # 干扰噪声比
Fs = 2000                    # 采样频率
T = 1/Fs                     # 采样时间
L = 2 * Fs                   # 信号长度
t = np.arange(0, L) * T      # 时间向量
P = 1                        # 噪声功率

# 生成白噪声
y = np.random.normal(0, np.sqrt(P), L)

# 绘制白噪声时域图
plt.figure()
plt.plot(1000 * t[:L//2], y[:L//2])
plt.title('White Noise')
plt.xlabel('time (milliseconds)')
plt.grid(True)
plt.show()

# 计算FFT
NFFT = 2**int(np.ceil(np.log2(L)))  # 下一个2的幂次
Y = np.fft.fft(y, NFFT) / L
f = Fs/2 * np.linspace(0, 1, NFFT//2 + 1)

# 绘制单边幅度谱
plt.figure()
plt.plot(f, 2 * 10 * np.log10(np.abs(Y[:NFFT//2 + 1])))
plt.title('Single-Sided Amplitude Spectrum of White Noise')
plt.xlabel('Frequency (Hz)')
plt.ylabel('|Y(f)|(dB)')
plt.grid(True)
plt.show()

# 带通滤波器设计
WI = 50  # 窄带窗口
FJ = 600  # 干扰频率
fcuts = [FJ - WI/2, FJ - WI/4, FJ + WI/4, FJ + WI/2]
mags = [0, 1, 0]
devs = [0.05, 0.05, 0.05]

# 计算Kaiser窗口参数
ripple_db = -20 * np.log10(np.min(devs))
width = min(fcuts[1] - fcuts[0], fcuts[3] - fcuts[2]) / (Fs/2)  # 归一化宽度
n, beta = kaiserord(ripple_db, width)

# 设计带通滤波器
wn = [fcuts[0]/(Fs/2), fcuts[3]/(Fs/2)]  # 归一化截止频率
h = firwin(n + 1, wn, window=('kaiser', beta), pass_zero=False)
Y_bp = lfilter(h, 1, y)  # 时域滤波结果

# 绘制滤波器频率响应 - 严格按照MATLAB的freqz格式
plt.figure()
w, h_freq = freqz(h, worN=8000)
# 上子图 - 幅度响应
plt.subplot(2, 1, 1)
plt.plot(w/np.pi, 20 * np.log10(np.abs(h_freq)))
plt.ylabel('Magnitude (dB)')
plt.grid(True)
# 下子图 - 相位响应
plt.subplot(2, 1, 2)
plt.plot(w/np.pi, np.angle(h_freq))
plt.xlabel('Normalized Frequency (×π rad/sample)')
plt.ylabel('Phase (radians)')
plt.grid(True)
plt.tight_layout()
plt.show()

# 绘制原始噪声和滤波后噪声
plt.figure()
plt.subplot(2, 1, 1)
plt.plot(1000 * t[:L//2], y[:L//2])
plt.title('White Noise')
plt.xlabel('time (milliseconds)')
plt.grid(True)

plt.subplot(2, 1, 2)
plt.plot(1000 * t[:L//2], Y_bp[:L//2])
plt.title('Narrow Band White Noise')
plt.xlabel('time (milliseconds)')
plt.grid(True)
plt.tight_layout()
plt.show()

# 绘制窄带噪声干扰时域图
plt.figure(figsize=(6, 3), facecolor='white')
plt.plot(1000 * t[:L//2], Y_bp[:L//2], color=[0, 0.4, 0.8])
plt.grid(True)
plt.title('Narrowband Noise jam')
plt.xlabel('time (milliseconds)')
plt.tight_layout()
plt.show()

# 计算滤波后信号的频谱
NFFT = 2**int(np.ceil(np.log2(L)))  # 下一个2的幂次
Y = np.fft.fft(Y_bp, NFFT) / L
f = Fs/2 * np.linspace(0, 1, NFFT//2 + 1)

# 绘制滤波后信号的频谱
plt.figure(figsize=(6, 3), facecolor='white')
plt.plot(f, 2 * 10 * np.log10(np.abs(Y[:NFFT//2 + 1])), color=[0, 0.4, 0.8])
plt.grid(True)
plt.title('Single-Sided Amplitude Spectrum of Narrowband Noise jam')
plt.xlabel('Frequency (Hz)')
plt.ylabel('|Y(f)|(dB)')
plt.tight_layout()
plt.show()
